Error: 'NoneType' object has no attribute 'startswith'

Hello,

In ovirt new version on Rocky 10.1 with dwh versions:

ovirt-engine-dwh-grafana-integration-setup-4.5.10-1.el10.noarch
ovirt-engine-dwh-setup-4.5.10-1.el10.noarch
ovirt-engine-dwh-4.5.10-1.el10.noarch

The serivce is stuck in 'activating' and i got the following error:

Jan 24 01:07:06 ovisp platform-python[14652]: 2026-01-24 01:07:06,394+0200 ovirt-engine-dwhd: ERROR run:572 Error: 'NoneType' object has no attribute 'startswith'

I found the 'startswith' function is used in the following files:
/usr/share/ovirt-engine/setup/plugins/ovirt-engine-rename/ovirt-engine-grafana-dwh/config.py (few times)
/usr/share/ovirt-engine-dwh/services/ovirt-engine-dwhd/ovirt-engine-dwhd.py

and i tried to mark the relevant lines but still when the script '/usr/share/ovirt-engine-dwh/services/ovirt-engine-dwhd/ovirt-engine-dwhd.py' running i got the following error:

[root@ovisp ~]# /usr/libexec/platform-python /usr/share/ovirt-engine-dwh/services/ovirt-engine-dwhd/ovirt-engine-dwhd.py --redirect-output --systemd=notify start
[root@ovisp ~]# 

(/var/log/messages)
Jan 24 01:13:16 ovisp platform-python[15782]: 2026-01-24 01:13:16,233+0200 ovirt-engine-dwhd: ERROR run:572 Error: 'NoneType' object has no attribute 'startswith'

Thanks,
Nir.

2026-01-26 12:58:11|ETL Service Started
ovirtEngineDbDriverClass|org.postgresql.Driver
ovirtEngineHistoryDbJdbcConnection|jdbc:postgresql://localhost:5432/ovirt_engine_history?sslfactory=org.postgresql.ssl.NonValidatingFactory
hoursToKeepDaily|0
hoursToKeepHourly|720
ovirtEngineDbPassword|**********************
runDeleteTime|3
ovirtEngineDbJdbcConnection|jdbc:postgresql://localhost:5432/engine?sslfactory=org.postgresql.ssl.NonValidatingFactory
runInterleave|60
limitRows|limit 1000
ovirtEngineHistoryDbUser|ovirt_engine_history
ovirtEngineDbUser|engine
deleteIncrement|10
timeBetweenErrorEvents|300000
hoursToKeepSamples|24
deleteMultiplier|1000
lastErrorSent|2011-07-03 12:46:47.000000
etlVersion|4.5.10
dwhAggregationDebug|false
dwhUuid|6a4762d6-4627-455d-85c7-4222e8c379b3
ovirtEngineHistoryDbDriverClass|org.postgresql.Driver
ovirtEngineHistoryDbPassword|**********************
Exception in thread "main" java.lang.NoClassDefFoundError: com/ongres/scram/common/stringprep/StringPreparation
	at org.postgresql.core.v3.ConnectionFactoryImpl.lambda$doAuthentication$4(ConnectionFactoryImpl.java:848)
	at org.postgresql.core.v3.AuthenticationPluginManager.withPassword(AuthenticationPluginManager.java:81)
	at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:835)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:203)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:258)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:54)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:263)
	at org.postgresql.Driver.makeConnection(Driver.java:444)
	at org.postgresql.Driver.connect(Driver.java:297)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:683)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:230)
	at routines.system.SharedDBConnection.getConnection(SharedDBConnection.java:63)
	at routines.system.SharedDBConnection.getDBConnection(SharedDBConnection.java:131)
	at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tJDBCConnection_6Process(HistoryETL.java:6184)
	at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tContextDump_1Process(HistoryETL.java:6095)
	at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tFileInputProperties_1Process(HistoryETL.java:5713)
	at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tRowGenerator_2Process(HistoryETL.java:5137)
	at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tLibraryLoad_1Process(HistoryETL.java:4773)
	at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tPrejob_1Process(HistoryETL.java:4662)
	at ovirt_engine_dwh.historyetl_4_5.HistoryETL.runJobInTOS(HistoryETL.java:11744)
	at ovirt_engine_dwh.historyetl_4_5.HistoryETL.main(HistoryETL.java:11391)
Caused by: java.lang.ClassNotFoundException: com.ongres.scram.common.stringprep.StringPreparation
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 21 more

[WORKAROUND]: see ath the end of this post
I have the same issue here. The 4.5.7 engine run with almalinux 10.1 (ovirt-engine-dwh-4.5.10-1.el10.noarch). I checked that the DB passwords (stored under /etc/ovirt-engine-dwh/ovirt-engine-dwhd.conf.d/10-setup-database.conf) are correct.
I tried also enable ongres-stringprep in /usr/share/ovirt-engine-dwh/bin/dwh-classpath.sh:
```
patch -p0 -d/ <<EOF
*** /usr/share/ovirt-engine-dwh/bin/dwh-classpath.sh.orig       2026-02-09 10:06:00.859000000 +0100
--- /usr/share/ovirt-engine-dwh/bin/dwh-classpath.sh    2026-02-09 10:33:58.397000000 +0100
***************
*** 63,70 ****
        [ -n "${jackson_annotations}" ] || die "Cannot find jackson-annotations"
        if [ "${what}" = "run" ]; then
                postgresql_jdbc="$(build-classpath postgresql-jdbc)" || die "Canot find postgreql-jdbc"
        fi
!       output="${output}:${dom4j}:${commons_collections}:${jackson_core}:${jackson_databind}:${jackson_annotations}:${postgresql_jdbc}"
  else
        die "Cannot find a method to acquire dependencies"
  fi
--- 63,72 ----
        [ -n "${jackson_annotations}" ] || die "Cannot find jackson-annotations"
        if [ "${what}" = "run" ]; then
                postgresql_jdbc="$(build-classpath postgresql-jdbc)" || die "Canot find postgreql-jdbc"
+               ongres_stringprep="$(build-classpath ongres-stringprep 2> /dev/null)"
+               [ -n "${ongres_stringprep}" ] || die "Cannot find ongres-stringprep"
        fi
!       output="${output}:${dom4j}:${commons_collections}:${jackson_core}:${jackson_databind}:${jackson_annotations}:${postgresql_jdbc}:${ongres_stringprep}"
  else
        die "Cannot find a method to acquire dependencies"
  fi
EOF
```

I launched using debug:
```
[ovirt@ovirt-engine ~]$ export OVIRT_SERVICE_DEBUG=1
[ovirt@ovirt-engine ~]$ /usr/share/ovirt-engine-dwh/services/ovirt-engine-dwhd/ovirt-engine-dwhd.py start
ovirt-engine-dwhd[87784] DEBUG _daemon:444 daemon entry pid=87784
ovirt-engine-dwhd[87784] DEBUG _daemon:445 background=False
ovirt-engine-dwhd[87784] DEBUG loadFile:63 loading config '/usr/share/ovirt-engine-dwh/services/ovirt-engine-dwhd/ovirt-engine-dwhd.conf'
ovirt-engine-dwhd[87784] DEBUG loadFile:63 loading config '/etc/ovirt-engine-dwh/ovirt-engine-dwhd.conf.d/10-setup-database.conf'
ovirt-engine-dwhd[87784] DEBUG loadFile:63 loading config '/etc/ovirt-engine-dwh/ovirt-engine-dwhd.conf.d/10-setup-grafana-database.conf'
ovirt-engine-dwhd[87784] DEBUG loadFile:63 loading config '/etc/ovirt-engine-dwh/ovirt-engine-dwhd.conf.d/10-setup-scale.conf'
ovirt-engine-dwhd[87784] DEBUG loadFile:63 loading config '/etc/ovirt-engine-dwh/ovirt-engine-dwhd.conf.d/10-setup-uuid.conf'
ovirt-engine-dwhd[87784] DEBUG getJavaHome:53 JAVA_HOME: /usr/lib/jvm/java-21-openjdk
ovirt-engine-dwhd[87784] DEBUG _clear:116 removing directory '/tmp/tmpjoheo6x_'
ovirt-engine-dwhd[87784] DEBUG _getClasspath:61 classpath: :/usr/share/ovirt-engine-wildfly/modules/system/layers/base/org/dom4j/main/dom4j-2.1.3.jar:/usr/share/java/apache-commons-collections.jar:/usr/share/java/jackson-core.jar:/usr/share/java/jackson-databind.jar:/usr/share/java/jackson-annotations.jar:/usr/share/java/postgresql-jdbc.jar:/usr/share/java/ongres-stringprep/saslprep.jar:/usr/share/java/ongres-stringprep/stringprep.jar
ovirt-engine-dwhd[87784] DEBUG _daemon:482 I am a daemon 87784
ovirt-engine-dwhd[87784] DEBUG _setLimits:422 Setting rlimits
ovirt-engine-dwhd[87784] DEBUG daemonAsExternalProcess:328 executing daemon: exe=/usr/lib/jvm/java-21-openjdk/bin/java, args=['ovirt-engine-dwhd', '-Dorg.ovirt.engine.dwh.settings=/tmp/tmpjoheo6x_/settings.properties', '-Xms1g', '-Xmx1g', '-classpath', '/usr/share/ovirt-engine-dwh/lib/*::/usr/share/ovirt-engine-wildfly/modules/system/layers/base/org/dom4j/main/dom4j-2.1.3.jar:/usr/share/java/apache-commons-collections.jar:/usr/share/java/jackson-core.jar:/usr/share/java/jackson-databind.jar:/usr/share/java/jackson-annotations.jar:/usr/share/java/postgresql-jdbc.jar:/usr/share/java/ongres-stringprep/saslprep.jar:/usr/share/java/ongres-stringprep/stringprep.jar', 'ovirt_engine_dwh.historyetl_4_5.HistoryETL', '--context=Default'], env={'SHELL': '/bin/bash', 'HISTCONTROL': 'ignoredups', 'HISTSIZE': '1000', 'HOSTNAME': 'ovirt-engine.ovirt', 'OVIRT_SERVICE_DEBUG': '1', 'GPG_TTY': '/dev/pts/0', 'PWD': '/var/lib/ovirt-engine', 'LOGNAME': 'ovirt', 'HOME': '/var/lib/ovirt-engine', 'LANG': 'en_US.UTF-8', 'LS_COLORS': 'rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=01;37;41:su=37;41:sg=30;43:ca=00:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.7z=01;31:*.ace=01;31:*.alz=01;31:*.apk=01;31:*.arc=01;31:*.arj=01;31:*.bz=01;31:*.bz2=01;31:*.cab=01;31:*.cpio=01;31:*.crate=01;31:*.deb=01;31:*.drpm=01;31:*.dwm=01;31:*.dz=01;31:*.ear=01;31:*.egg=01;31:*.esd=01;31:*.gz=01;31:*.jar=01;31:*.lha=01;31:*.lrz=01;31:*.lz=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.lzo=01;31:*.pyz=01;31:*.rar=01;31:*.rpm=01;31:*.rz=01;31:*.sar=01;31:*.swm=01;31:*.t7z=01;31:*.tar=01;31:*.taz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tgz=01;31:*.tlz=01;31:*.txz=01;31:*.tz=01;31:*.tzo=01;31:*.tzst=01;31:*.udeb=01;31:*.war=01;31:*.whl=01;31:*.wim=01;31:*.xz=01;31:*.z=01;31:*.zip=01;31:*.zoo=01;31:*.zst=01;31:*.avif=01;35:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=01;36:*.au=01;36:*.flac=01;36:*.m4a=01;36:*.mid=01;36:*.midi=01;36:*.mka=01;36:*.mp3=01;36:*.mpc=01;36:*.ogg=01;36:*.ra=01;36:*.wav=01;36:*.oga=01;36:*.opus=01;36:*.spx=01;36:*.xspf=01;36:*~=00;90:*#=00;90:*.bak=00;90:*.crdownload=00;90:*.dpkg-dist=00;90:*.dpkg-new=00;90:*.dpkg-old=00;90:*.dpkg-tmp=00;90:*.old=00;90:*.orig=00;90:*.part=00;90:*.rej=00;90:*.rpmnew=00;90:*.rpmorig=00;90:*.rpmsave=00;90:*.swp=00;90:*.tmp=00;90:*.ucf-dist=00;90:*.ucf-new=00;90:*.ucf-old=00;90:', 'TERM': 'xterm-256color', 'LESSOPEN': '||/usr/bin/lesspipe.sh %s', 'USER': 'ovirt', 'SHLVL': '1', 'DEBUGINFOD_IMA_CERT_PATH': '/etc/keys/ima:', 'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin', 'MAIL': '/var/spool/mail/ovirt', '_': '/usr/share/ovirt-engine-dwh/services/ovirt-engine-dwhd/ovirt-engine-dwhd.py', 'LC_ALL': 'en_US.UTF-8'}
ovirt-engine-dwhd[87784] DEBUG daemonAsExternalProcess:336 creating process
ovirt-engine-dwhd[87784] DEBUG daemonAsExternalProcess:344 waiting for termination of pid=87910
2026-02-09 10:40:34|ETL Service Started
ovirtEngineDbDriverClass|org.postgresql.Driver
ovirtEngineHistoryDbJdbcConnection|jdbc:postgresql://localhost:5432/ovirt_engine_history?sslfactory=org.postgresql.ssl.NonValidatingFactory
hoursToKeepDaily|0
hoursToKeepHourly|720
ovirtEngineDbPassword|**********************
runDeleteTime|3
ovirtEngineDbJdbcConnection|jdbc:postgresql://localhost:5432/engine?sslfactory=org.postgresql.ssl.NonValidatingFactory
runInterleave|60
limitRows|limit 1000
ovirtEngineHistoryDbUser|ovirt_engine_history
ovirtEngineDbUser|engine
deleteIncrement|10
timeBetweenErrorEvents|300000
hoursToKeepSamples|24
deleteMultiplier|1000
lastErrorSent|2011-07-03 12:46:47.000000
etlVersion|4.5.10
dwhAggregationDebug|false
dwhUuid|93c4fc7f-74ef-43af-8529-3de33f2f763d
ovirtEngineHistoryDbDriverClass|org.postgresql.Driver
ovirtEngineHistoryDbPassword|**********************
Exception in thread "main" java.lang.NoClassDefFoundError: com/ongres/scram/common/stringprep/StringPreparation
        at org.postgresql.core.v3.ConnectionFactoryImpl.lambda$doAuthentication$4(ConnectionFactoryImpl.java:848)
        at org.postgresql.core.v3.AuthenticationPluginManager.withPassword(AuthenticationPluginManager.java:81)
        at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:835)
        at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:203)
        at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:258)
        at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:54)
        at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:263)
        at org.postgresql.Driver.makeConnection(Driver.java:444)
        at org.postgresql.Driver.connect(Driver.java:297)
        at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:683)
        at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:230)
        at routines.system.SharedDBConnection.getConnection(SharedDBConnection.java:63)
        at routines.system.SharedDBConnection.getDBConnection(SharedDBConnection.java:131)
        at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tJDBCConnection_6Process(HistoryETL.java:6184)
        at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tContextDump_1Process(HistoryETL.java:6095)
        at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tFileInputProperties_1Process(HistoryETL.java:5713)
        at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tRowGenerator_2Process(HistoryETL.java:5137)
        at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tLibraryLoad_1Process(HistoryETL.java:4773)
        at ovirt_engine_dwh.historyetl_4_5.HistoryETL.tPrejob_1Process(HistoryETL.java:4662)
        at ovirt_engine_dwh.historyetl_4_5.HistoryETL.runJobInTOS(HistoryETL.java:11744)
        at ovirt_engine_dwh.historyetl_4_5.HistoryETL.main(HistoryETL.java:11391)
Caused by: java.lang.ClassNotFoundException: com.ongres.scram.common.stringprep.StringPreparation
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
        ... 21 more
ovirt-engine-dwhd[87784] DEBUG daemonAsExternalProcess:349 terminated pid=87910 rc=1
ovirt-engine-dwhd[87784] DEBUG _clear:116 removing directory '/tmp/tmpjoheo6x_'
ovirt-engine-dwhd[87784] ERROR run:572 Error: process terminated with status code 1
ovirt-engine-dwhd[87784] DEBUG run:577 exception
Traceback (most recent call last):
  File "/usr/lib/python3.12/site-packages/ovirt_engine/service.py", line 570, in run
    self._daemon()
  File "/usr/lib/python3.12/site-packages/ovirt_engine/service.py", line 488, in _daemon
    self.daemonContext()
  File "/usr/share/ovirt-engine-dwh/services/ovirt-engine-dwhd/ovirt-engine-dwhd.py", line 231, in daemonContext
    self.daemonAsExternalProcess(
  File "/usr/lib/python3.12/site-packages/ovirt_engine/service.py", line 356, in daemonAsExternalProcess
    raise RuntimeError(
RuntimeError: process terminated with status code 1
```

The problem is that in the /usr/share/java/ongres-stringprep/stringprep.jar There ist only the StringPrep.class not StringPreparation

[WORKAROUND] Finally, I downloaded https://jdbc.postgresql.org/download/postgresql-42.7.9.jar and put into /usr/share/ovirt-engine-dwh/lib/. Now dwhd is running

--- a/packaging/bin/dwh-classpath.sh
+++ b/packaging/bin/dwh-classpath.sh
@@ -47,10 +47,15 @@ if [ -x /usr/bin/build-classpath ]; then
 	jackson_annotations="$(build-classpath jackson-annotations 2> /dev/null)"
 	[ -z "${jackson_annotations}" ] && jackson_annotations="$(build-classpath jackson-annotations 2> /dev/null)"
 	[ -n "${jackson_annotations}" ] || die "Cannot find jackson-annotations"
+	output="${output}:${dom4j}:${commons_collections}:${jackson_core}:${jackson_databind}:${jackson_annotations}"
 	if [ "${what}" = "run" ]; then
-		postgresql_jdbc="$(build-classpath postgresql-jdbc)" || die "Canot find postgreql-jdbc"
+		postgresql_jdbc="$(build-classpath postgresql-jdbc)" || die "Cannot find postgreql-jdbc"
+		output="${output}:${postgresql_jdbc}"
+		ongres_scram="$(build-classpath ongres-scram)"
+		[ -n "${ongres_scram}" ] && output="${output}:${ongres_scram}"
+		ongres_stringprep="$(build-classpath ongres-stringprep)"
+		[ -n "${ongres_stringprep}" ] && output="${output}:${ongres_stringprep}"
 	fi
-	output="${output}:${dom4j}:${commons_collections}:${jackson_core}:${jackson_databind}:${jackson_annotations}:${postgresql_jdbc}"
 else
 	die "Cannot find a method to acquire dependencies"
 fi

--- a/packaging/bin/dwh-classpath.sh
+++ b/packaging/bin/dwh-classpath.sh
@@ -31,21 +31,7 @@ if [ -z "${WILDFLYJAVACONF}" ]; then
 	WILDFLYJAVACONFDIR="$(readlink -f $(dirname $(dirname $0)))/wildflyjavaconf"
 fi
 
-if [ -x /usr/bin/java-config ]; then
-	PACKAGES_BUILD="dom4j-1 commons-collections jackson-core jackson-databind jackson-annotations"
-	PACKAGES_RUNTIME="jdbc-postgresql"
-
-	packages="${PACKAGES_BUILD}"
-	args=
-	if [ "${what}" = "run" ]; then
-		packages="${packages} ${PACKAGES_RUNTIME}"
-		args="${args} -d"
-	fi
-	for package in ${packages}; do
-		output="${output}:$(java-config ${args} -p ${package} 2> /dev/null)" \
-			|| die "Cannot locate ${package}"
-	done
-elif [ -x /usr/bin/build-classpath ]; then
+if [ -x /usr/bin/build-classpath ]; then
 	dom4j="$(build-classpath dom4j 2> /dev/null)"
 	[ -z "${dom4j}" ] && dom4j="$(JAVACONFDIRS=${WILDFLYJAVACONFDIR} build-classpath dom4j 2> /dev/null)"
 	[ -n "${dom4j}" ] || die "Cannot find dom4j"

